include/irr/asset/material_compiler/IR.h

```diff
@@ -0,0 +1,429 @@
+        {
+            opacity.source = EPS_CONSTANT;
+            opacity.value.constant = color_t(1.f);
+        }
+
+        SParameter<color_t> opacity;
+        bool thin;
+    };
+
+    struct CGeomModifierNode : public INode
+    {
+        enum E_TYPE
+        {
+            ET_DISPLACEMENT,
+            ET_HEIGHT,
+            ET_NORMAL,
+            ET_DERIVATIVE
+        };
+        enum E_SOURCE
+        {
+            ESRC_UV_FUNCTION,
+            ESRC_TEXTURE
+        };
+
+        CGeomModifierNode(E_TYPE t) : INode(ES_GEOM_MODIFIER), type(t) {}
+
+        E_TYPE type;
+        E_SOURCE source;
+        //TODO some union member for when source==ESRC_UV_FUNCTION, no idea what type
+        //in fact when we want to translate MDL function of (u,v) into this IR we could just create an image being a 2D plot of this function with some reasonable quantization (pixel dimensions)
+        //union {
+            STextureSource texture;
+        //};
+    };
+
+    struct CEmissionNode : INode
+    {
+        CEmissionNode() : INode(ES_EMISSION) {}
+
+        float intensity;
+    };
+
+    struct CBSDFCombinerNode : INode
+    {
+        enum E_TYPE
+        {
+            //mix of N BSDFs
+            ET_MIX,
+            //blend of 2 BSDFs weighted by constant or texture
+            ET_WEIGHT_BLEND,
+            //blend of 2 BSDFs weighted by fresnel
+            ET_FRESNEL_BLEND,
+            //blend of 2 BSDFs weighted by custom direction-based curve
+            ET_CUSTOM_CURVE_BLEND,
+            ET_DIFFUSE_AND_SPECULAR
+        };
+
+        E_TYPE type;
+
+        CBSDFCombinerNode(E_TYPE t) : INode(ES_BSDF_COMBINER), type(t) {}
+    };
+    struct CBSDFBlendNode : CBSDFCombinerNode
+    {
+        CBSDFBlendNode() : CBSDFCombinerNode(ET_WEIGHT_BLEND) {}
+
+        SParameter<float> weight;
+    };
+    struct CBSDFMixNode : CBSDFCombinerNode
+    {
+        CBSDFMixNode() : CBSDFCombinerNode(ET_MIX) {}
+
+        float weights[MAX_CHILDREN];
+    };
+
+    struct CBSDFNode : INode
+    {
+        enum E_TYPE
+        {
+            ET_DIFFTRANS,
+            ET_SPECULAR_DELTA,
+            ET_MICROFACET_DIFFUSE,
+            ET_MICROFACET_SPECULAR,
+            ET_SHEEN,
+            ET_COATING
+        };
+
+        CBSDFNode(E_TYPE t) :
+            INode(ES_BSDF),
+            type(t),
+            eta(1.f,1.f,1.f),
+            etaK(0.f,0.f,0.f)
+        {}
+
+        E_TYPE type;
+        color_t eta, etaK;
+    };
+    struct CMicrofacetSpecularBSDFNode : CBSDFNode
+    {
+        enum E_NDF
+        {
+            ENDF_BECKMANN,
+            ENDF_GGX,
+            ENDF_ASHIKHMIN_SHIRLEY,
+            ENDF_PHONG
+        };
+        enum E_SHADOWING_TERM
+        {
+            EST_SMITH,
+            EST_VCAVITIES
+        };
+        enum E_SCATTER_MODE
+        {
+            ESM_REFLECT,
+            ESM_TRANSMIT
+        };
+
+        CMicrofacetSpecularBSDFNode(E_TYPE t = ET_MICROFACET_SPECULAR) : CBSDFNode(t) {}
```

Comment:
why is the type even configurable?

Reply:
becase some other node type inherits from this iirc

Reply:
which node?

Reply:
coating

Reply:
make this constructor protected, dont provide default arg, and make it plain as day that its only coating that overrides it